Resolving comments from Bool Tensor for CPU PR

izdeby opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #18166 Bool Tensor for CUDA
* **#18165 Resolved comments from Bool Tensor for CPU PR**
-------
------------
This is a follow up PR that resolves some additional feedback on one the of previous Bool Tensor PRs.

@gchanan, here is a list of almost all the comments from the original PR with respective fixes and replies:

**[utils/python_scalars.h]** why is this converting from uint8_t and not bool? (comment?)
When i was adding this, i was testing by creating a tensor and then calling its .tolist(). it worked for bool and uint8_t equally good so i left uint8_t as thought it makes more sense as we are calling PyBool_FromLong.  Changing it to bool.

**[ATen/Dispatch.h]**better name?.
fixed.

**[test/test_torch.py]** what about other factories, such as full? (and more).
There is a test that goes through the factory methods - test_tensor_factories_empty. i added some bool cases above it and added a comment that once CUDA will be done, i will unite them and it will iterate not just between CUDA and CPU but also all types.   Adding all bool cases now. Will unite in CUDA PR.

**[generic/THTensorMath.h]** any changes in this file actually needed?
Bad merge. Fixed.

**[TH/THTensor.h]** this generates code for random, clampedRandom, and cappedRandom -- do we have tests for all of these with bool?
Added

**[c10/core/ScalarType.h]** I'm not very confident about the lack of Bool here -- can you look at the call sites and see what makes sense to do here?
Added bool to the macro and created a similar one without for a single case which fails the build with errors:

_./torch/csrc/jit/symbolic_variable.h:79:20: error: ambiguous overload for ‘operator*’ (operand types are ‘const torch::jit::SymbolicVariable’ and ‘torch::jit::Value*’)
return (*this) * insertConstant(rhs);_

Differential Revision: [D14605105](https://our.internmc.facebook.com/intern/diff/D14605105)

izdeby committed:
Resolved comments from Bool Tensor for CPU PR

izdeby committed:
Update on "[WIP]Resolving comments from Bool Tensor for CPU PR"

izdeby commented:
@gchanan, ready for review. Couple moments: 
1. Moving ScalarTypeToCType to details namespace didn't work for me. Renaming struct didn't help, but renaming namespace did. will follow up on this in a separate PR if thats ok.
2. using AT_FORALL_SCALAR_TYPES( with no bool) in c10/core/Scalar.h as discussed offline. 
3. Will kill random_geometric in a separate PR to keep this PR cleaner.

ezyang commented on test/common_utils.py:
```diff
@@ -404,11 +404,15 @@ def assertTensorsEqual(a, b):
+                        if a.dtype == torch.bool and b.dtype == torch.bool:
+                            self.assertEqual(a.tolist(), b.tolist())
+                            a = torch.as_tensor(a, dtype=torch.int)
+                            b = torch.as_tensor(b, dtype=torch.int)
```
This is pretty un-expected.  Why should we convert into an int? Haven't we already tested that the contents of the two tensors are equal?

gchanan replied:
I think this would make sense if you just got rid of the first assertEqual.  It's also clearer if you just used `to` instead of `as_tensor` -- you already know it's a tensor.  And probably want a comment about why you want to do this (because subtraction isn't currently supported on bool).

ezyang commented on aten/src/ATen/native/TensorFactories.cpp:
```diff
@@ -141,7 +141,7 @@ Tensor& empty_out(Tensor& result, IntArrayRef size) {
   }
 
-AT_FORALL_SCALAR_TYPES(DEFINE_CAST_OP)
+AT_FORALL_SCALAR_TYPES_AND_BOOL(DEFINE_CAST_OP)
```
I wonder if it's possible to adjust our `AT_FORALL` macros in the same way we adjusted the `AT_DISPATCH` macros. Maybe this is not conveniently possible.

gchanan replied:
I think we can -- we don't need to worry about backwards incompatibility either!

izdeby replied:
will fix as a separate PR

ezyang commented on test/test_torch.py:
```diff
@@ -2901,18 +2901,58 @@ def test_tensor_factory_cuda_type(self):
         self.assertTrue(x.is_cuda)
         torch.set_default_tensor_type(saved_type)
 
+    def test_tensor_randoms_bool(self):
```
I feel a little uncomfortable about the fact that we are not actually testing against the most likely bug, which is that we are generating a number uniformly at random from 0-255 and then casting it to bool (which means that true would be 255x more likely to show up than false). We're not testing this because you never actually look at the values in question.

I'd suggest determining some acceptable level of flakiness, and then explicitly testing the distribution up to some p-value. As we seed before all tests, the test will deterministically pass or fail (unless someone perturbs the test itself; thus the acceptable level of flakiness.) An example is in this test:

```
    def test_random(self):
        # This test is flaky with p<=(2/(ub-lb))^200=6e-36
        t = torch.FloatTensor(200)
        lb = 1
        ub = 4

        t.fill_(-1)
        t.random_(lb, ub)
        self.assertEqual(t.min(), lb)
        self.assertEqual(t.max(), ub - 1)

        t.fill_(-1)
        t.random_(ub)
        self.assertEqual(t.min(), 0)
        self.assertEqual(t.max(), ub - 1)
```

izdeby replied:
Ok, removing bool support for all randoms except the randint.

ezyang reviewed (approved):
Accepted with comments

gchanan commented:
On the situation:
> ./torch/csrc/jit/symbolic_variable.h:79:20: error: ambiguous overload for ‘operator*’ (operand types are ‘const torch::jit::SymbolicVariable’ and ‘torch::jit::Value*’)
return (*this) * insertConstant(rhs);

This is because `Value*` is both implicitly convertible to `SymbolicVariable` and `bool`.  As @ezyang points out, if you declare the bool constructor using `std::enable_if` you can still have a bool constructor and avoid ambiguous overloads because template specialization doesn't involve implicit conversions.

izdeby commented on aten/src/ATen/Declarations.cwrap:
```diff
@@ -2573,10 +2573,12 @@
     - CPU
     - CUDA
   cname: uniform
+  cpu_bool: True
```
_uniform_ is being used for rand()
https://github.com/pytorch/pytorch/blob/670f509984efce8aab0b0b5b166ff01472af6f1c/aten/src/ATen/native/TensorFactories.cpp#L277

removing this as we will support only randint for bool.

gchanan replied:
But why do we need support for `rand`?  It's not supported for other integral types.

gchanan commented on aten/src/TH/generic/THTensorMath.h:
```diff
@@ -22,6 +22,7 @@ TH_API void THTensor_(scatterFill)(THTensor *tensor, int dim, THLongTensor *inde
 
 TH_API accreal THTensor_(dot)(THTensor *t, THTensor *src);
 
+TH_API scalar_t THTensor_(powOne)(scalar_t x, scalar_t y);
```
That's because you didn't move `powOne` before the first usage in THTensorMath.cpp.  It's not used outside of that file, it doesn't need to be exposed in a .h file.

izdeby commented:
@vadimkantorov, yes its not implemented yet. So far i've finished the CPU part and now working on the CUDA changes (next PR in the current stack). Once its done, i will implement indexing and ops. If everything will go according to the schedule, it should be done by the end of march.

izdeby committed:
Update on "Resolving comments from Bool Tensor for CPU PR"

izdeby committed:
Update on "[wip]Resolving comments from Bool Tensor for CPU PR"

izdeby commented:
Hey @gchanan and @ezyang for the review and suggestions on how to resolve some issues. I've updated the PR in order to resolve some additional feedback.

Fix: 
Moved ScalarTypeToCType to detail namespace using the workaround and added a comment
Removed support to _uniform_ for bool tensor. This will disable most of randoms for bool and only randint will be available.
Removed powOne declaration.
Added bool constructor for Scalar. 

gchanan commented on aten/src/ATen/Dispatch.h:
```diff
@@ -12,6 +12,25 @@
+  using type = at::Half;
+
+  // This is a workaround the CUDA bug which prevents ::detail::ScalarTypeToCType to be resolved.
+  static at::Half t;
```
nit: can you link to an explanation of the issue?

gchanan commented on aten/src/ATen/Dispatch.h:
```diff
@@ -119,26 +138,6 @@ inline void deprecated_AT_DISPATCH_ALL_TYPES_AND_HALF_AND_COMPLEX() {}
     }                                                                        \
   }()
 
-#define AT_DISPATCH_ALL_TYPES_AND_HALF(TYPE, NAME, ...)                      \
```
why are you deleting AT_DISPATCH_ALL_TYPES_AND_HALF?  We want to maintain this for backwards compatibility, see https://github.com/pytorch/pytorch/pull/17996.

izdeby replied:
I didn't know about the backwards compatibility issue. I deleted it as we decided to replace some of these macros with  AT_DISPATCH_ALL_TYPES_AND(SCALARTYPE, TYPE, NAME, ...).
Reverting.

gchanan commented on c10/core/Scalar.h:
```diff
@@ -35,6 +35,12 @@ class C10_API Scalar {
 #undef DEFINE_IMPLICIT_CTOR
 
+template<typename T>
+Scalar(typename std::enable_if<std::is_same<T, bool>::type>::type, bool vv)
```
does this work?  Have you tried calling it this way in C++?

Look at the examples from https://en.cppreference.com/w/cpp/types/enable_if -- they put the `enable_if` in the template.

izdeby replied:
Fixed and tested [via](https://github.com/izdeby/playground/blob/dedf311e866d4b124081e881c522e32dab86123d/C%2B%2B/enableIfTest.cpp).

izdeby committed:
Update on "Resolving comments from Bool Tensor for CPU PR"

izdeby committed:
Update on "Resolving comments from Bool Tensor for CPU PR"

izdeby committed:
Update on "Resolving comments from Bool Tensor for CPU PR"

gchanan commented on aten/src/ATen/Dispatch.h:
```diff
@@ -12,6 +12,29 @@
+struct ScalarTypeToCType<at::ScalarType::Half> {
+  using type = at::Half;
+
+  // This is a workaround for the CUDA bug which prevents ::detail::ScalarTypeToCType<T>::type being used directly
```
nit: you should reference an actual filed issue.  Like one on github or a gist.

gchanan commented on c10/core/Scalar.h:
```diff
@@ -35,6 +35,13 @@ class C10_API Scalar {
 
 #undef DEFINE_IMPLICIT_CTOR
 
+template <typename T,
```
comment.

izdeby committed:
Update on "Resolving comments from Bool Tensor for CPU PR"

izdeby committed:
Update on "Resolving comments from Bool Tensor for CPU PR"
